File: Medium/1208-Get Equal Substrings Within Budget/Two Pointers.rb
# @param {String} s
# @param {String} t
# @param {Integer} max_cost
# @return {Integer}
def equal_substring(s, t, max_cost)
  s = s.bytes
  t = t.bytes
  i = -1
  cost = 0
  ret = 0

  (0...s.size).each do |j|
    cost += (s[j] - t[j]).abs
    while cost > max_cost
      i += 1
      cost -= (s[i] - t[i]).abs
    end
    ret = [ret, j - i].max
  end

  ret
end

File: Problemset/1208-Get Equal Substrings Within Budget/Two Pointers.rs
impl Solution {
    pub fn equal_substring(s: String, t: String, max_cost: i32) -> i32 {
        let s = s.as_bytes();
        let t = t.as_bytes();
        let mut i = -1;
        let mut cost = 0;
        let mut ret = 0;

        for j in 0..s.len() {
            cost += (s[j] as i32 - t[j] as i32).abs();
            while cost > max_cost {
                i += 1;
                cost -= (s[i as usize] as i32 - t[i as usize] as i32).abs();
            }
            ret = ret.max(j as i32 - i);
        }

        ret
    }
}
